feat: allow customization of the HttpTransferCache.

JeanMeche opened the pull request:
`withHttpTransferCacheOptions()` accepts a set of options to allow customization of the cache :
* `disable` replaces `withNoHttpTransferCache()`
*  `exclude`: a callback to determine wether a request should be cached or not

BREAKING CHANGE: `withNoHttpTransferCache` is replaced by `withHttpTransferCacheOptions({ disable: true })`

Fixes #...

AndrewKushnir commented on packages/common/http/src/client.ts:
```diff
@@ -138,6 +139,7 @@ export class HttpClient {
           {[param: string]: string | number | boolean | ReadonlyArray<string|number|boolean>},
     reportProgress?: boolean, responseType: 'arraybuffer',
     withCredentials?: boolean,
+    excludeFromHttpTransfer?: boolean,
```
I think it'd be great to use "cache" in the name (or the entire `HttpTransferCache`):
```suggestion
    excludeFromHttpTransferCache?: boolean,
```
May be something like this?
```suggestion
    skipHttpTransferCache?: boolean,
```

AndrewKushnir replied:
"skip" is shorter, but "excludeFrom" has an interplay with the "exclude" option from the "withHttpTransferCache" function.

AndrewKushnir commented on packages/common/http/src/client.ts:
```diff
@@ -523,6 +540,14 @@ export class HttpClient {
         }
       }
 
+      // Should we inject(Console) ?
```
Yes :)

AndrewKushnir commented on packages/platform-browser/src/hydration.ts:
```diff
@@ -73,24 +68,28 @@ function hydrationFeature<FeatureKind extends HydrationFeatureKind>(
-  // This feature has no providers and acts as a flag that turns off
-  // HTTP transfer cache (which otherwise is turned on by default).
-  return hydrationFeature(HydrationFeatureKind.NoHttpTransferCache);
+export function withHttpTransferCacheOptions(
```
I think we should keep the `withNoHttpTransferCache()` function and rename this function to `withHttpTransferCache({ ... } /* options */)`.

alan-agius4 commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -24,22 +24,30 @@ interface TransferHttpResponse {
     req: HttpRequest<unknown>, next: HttpHandlerFn): Observable<HttpEvent<unknown>> {
-  const {isCacheActive} = inject(CACHE_STATE);
+  const {isCacheActive, exclude} = inject(CACHE_OPTIONS);
+  const skipCacheForRequest = exclude?.(req) === true || req.excludeFromHttpTransfer === true ||
```
```suggestion
  const skipCacheForRequest = exclude?.(req) || req.excludeFromHttpTransfer ||
```

JeanMeche committed:
feat: allow customization of the HttpTransferCache.

`withHttpTransferCache()` accepts a set of options to allow customization of the cache :
*  `exclude`: a callback to determine wether a request should be cached or not
* more to come...

Fixes #50117
